File: smash-ranks-android/app/src/main/java/com/garpr/android/misc/FavoritePlayersManager.java
package com.garpr.android.misc;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.garpr.android.models.AbsPlayer;
import com.garpr.android.models.FavoritePlayer;
import com.garpr.android.models.Region;

import java.util.List;

public interface FavoritePlayersManager {

    void addListener(@NonNull final OnFavoritePlayersChangeListener listener);

    void addPlayer(@NonNull final AbsPlayer player, @NonNull final Region region);

    void clear();

    boolean containsPlayer(@NonNull final AbsPlayer player);

    boolean containsPlayer(@NonNull final String playerId);

    @Nullable
    List<AbsPlayer> getAbsPlayers();

    @Nullable
    List<FavoritePlayer> getPlayers();

    boolean isEmpty();

    void removeListener(@Nullable final OnFavoritePlayersChangeListener listener);

    void removePlayer(@NonNull final AbsPlayer player);

    void removePlayer(@NonNull final String playerId);

    boolean showAddOrRemovePlayerDialog(@NonNull Context context,
            @Nullable final AbsPlayer player, @NonNull final Region region);

    int size();


    interface OnFavoritePlayersChangeListener {
        void onFavoritePlayersChanged(final FavoritePlayersManager manager);
    }

}

File: smash-ranks-android/app/src/main/java/com/garpr/android/misc/FavoritePlayersManager.kt
package com.garpr.android.misc

import android.content.Context

import com.garpr.android.models.AbsPlayer
import com.garpr.android.models.FavoritePlayer
import com.garpr.android.models.LiteRegion

interface FavoritePlayersManager {

    interface OnFavoritePlayersChangeListener {
        fun onFavoritePlayersChanged(manager: FavoritePlayersManager)
    }

    val absPlayers: List<AbsPlayer>?

    fun addListener(listener: OnFavoritePlayersChangeListener)

    fun addPlayer(player: AbsPlayer, region: LiteRegion)

    fun clear()

    operator fun contains(player: AbsPlayer): Boolean

    operator fun contains(playerId: String): Boolean

    val isEmpty: Boolean

    val players: List<FavoritePlayer>?

    fun removeListener(listener: OnFavoritePlayersChangeListener?)

    fun removePlayer(player: AbsPlayer)

    fun removePlayer(playerId: String)

    fun showAddOrRemovePlayerDialog(context: Context, player: AbsPlayer?, region: LiteRegion): Boolean

    val size: Int

}
